nixos/modules/services/misc/matrix-conduit.nix

```diff
@@ -0,0 +1,168 @@
+              default = [ "matrix.org" ];
+              description = "Servers trusted with signing server keys.";
+            };
+
+            global.address = mkOption {
+              type = types.str;
+              default = "127.0.0.1";
+              description = "Address to listen on for connections by the reverse proxy/tls terminator.";
+            };
+          };
+        };
+        default = {};
+        description = ''
+            generates the conduit.toml configuration file. Refer to
+            <link xlink:href="https://gitlab.com/famedly/conduit/-/blob/master/conduit-example.toml"/>
+            for details on supported values.
+            Note that database_path should not be set because the services reliance on systemd StateDir.
+        '';
+      };
+    };
+
+
+    config = mkIf cfg.enable {
+      services.conduit.settings = {
+        global.database_path = "/var/lib/matrix-conduit/conduit.db";
+      };
+      assertions = [
+        {
+          assertion = cfg.settings.global.server_name != null;
+          message = "services.conduit.settings.global.server_name must be set.";
+        }
+        {
+          assertion = cfg.settings.global.allow_federation -> cfg.iHaveReadTheFederationWarning;
+          message = ''
+            Upstream documentation contains the following warning:
+            You should not join Matrix rooms without asking the admins first.
+            We do not know whether Conduit is safe for general use yet, so you
+            should assume there is some chance that it breaks rooms permanently
+            for all participating users. We are not aware of such a bug today,
+            but we would like to do more testing.
+            You need to acknowledge this warning by setting
+            services.conduit.iHaveReadTheFederationWarning = true
+          '';
+        }
+      ];
+
+      systemd.services.conduit = {
+        description = "Conduit Matrix Server";
+        documentation = [ "https://gitlab.com/famedly/conduit/" ];
+        wantedBy = [ "multi-user.target" ];
+        environment = {
+          CONDUIT_CONFIG = configFile;
+        };
+        serviceConfig = {
+          DynamicUser = true;
+          LockPersonality = true;
+          MemoryDenyWriteExecute = true;
+          ProtectClock = true;
+          ProtectControlGroups = true;
+          ProtectHostname = true;
+          ProtectKernelLogs = true;
+          ProtectKernelModules = true;
+          ProtectKernelTunables = true;
+          PrivateDevices = true;
+          PrivateMounts = true;
+          PrivateUsers = true;
+          RestrictAddressFamilies = "AF_INET AF_INET6";
+          RestrictNamespaces = true;
+          RestrictRealtime = true;
+          SystemCallArchitectures = "native";
+          SystemCallFilter = "@system-service";
+          StateDirectory = "matrix-conduit";
+          ExecStart="${cfg.package}/bin/conduit";
+          Restart="always";
```

Comment:
Does it crash often? Why are we always restarting? Should each crash be a bug report for upstream instead and noticed by us (you)?

Reply:
Just as a general policy, definitely not specific to this software... gosh I hate when I'm called at 3AM because a service crashed and didn't automatically start back up.

I think @peterhoeg has had similar thoughts on this...

Reply:
A bit more from my point of view. Server software is generally meant to run. Often functional uptime is the most important metric. No, not the only unpretentious metric, but usually the most important. I want to know when services have crashed, but this is what I use monitoring solutions for. You can use some full blown monitoring solution and get loads of useful data, or you can go as simple as using the `OnFailure` mechanism of `systemd`.

There isn't a universal right or wrong answer here, but I thought I would share my opinion on the topic because it has come up a few times.

Reply:
It's from their upstream service file. I fixed it to more reasonable values from... another upstream service file.

Reply:
@aanderse @andir Are you now both satisfied with the compromise?

Reply:
I was just commenting in general, so it isn't important if I'm satisfied here :smile: 

But yes, that value satisfies me :+1:

Reply:
`on-failure` is absolutely fine with me even thought I'd prefer if it fails loud and clear when it fails.

If services fail that is a bug that should be reported and `Restart=...` policies should IMHO only be set if that becomes an actual issue that is recurring.

Reply:
Failures need not necessarily be caused by a bug. For example, temporarily running out of resources may be. I find simply restarting the application acceptable, if it gracefully supports it. And I wouldn't necessarily consider it a bug in the code.